Provisioning: Use dashboard checksum field as change indicator

cristi- opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR helps to ensure provisioned dashboards are reloaded even if the file mtime hasn't changed but it's checksum has.

**Which issue(s) this PR fixes**:

https://github.com/grafana/grafana/issues/29796
<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=29797) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=29797) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=29797) it.</sub>

wbrowne reviewed (commented):
Thanks for the contribution @cristi-! 😄  It would be great if you could add a test for this and then it should be good to go

cristi- committed:
Add tests for checksum validation

Check dashboard checksum for provisioned dashboards

Remove extra error

Remove extra error

Ignore mtime when checking privisioned dashboards; Fix tests

Fix linting errors

aknuds1 commented on pkg/services/provisioning/dashboards/file_reader_test.go:
```diff
@@ -148,6 +148,47 @@ func TestDashboardFileReader(t *testing.T) {
+				err = reader.startWalkingDisk()
+				So(err, ShouldBeNil)
+				So(len(fakeService.inserted), ShouldEqual, 1)
+			})
```
Should be a blank line between tests.
```suggestion
			})

```

aknuds1 commented on pkg/services/provisioning/dashboards/file_reader_test.go:
```diff
@@ -148,6 +148,47 @@ func TestDashboardFileReader(t *testing.T) {
 
+			Convey("Older dashboard file with the same checksum will not replace old dashboard", func() {
+				cfg.Options["path"] = oneDashboard
+				absPath, _ := filepath.Abs(oneDashboard + "/dashboard1.json")
+				stat, _ := os.Stat(oneDashboard + "/dashboard1.json")
+				file, _ := os.Open(absPath)
+				defer file.Close()
+
+				checksum, _ := util.Md5Sum(file)
```
```suggestion
				absPath, err := filepath.Abs(oneDashboard + "/dashboard1.json")
				So(err, ShouldBeNil)
				stat, err := os.Stat(oneDashboard + "/dashboard1.json")
				So(err, ShouldBeNil)
				file, err := os.Open(absPath)
				So(err, ShouldBeNil)
				defer file.Close()

				checksum, _ := util.Md5Sum(file)
```

aknuds1 commented on pkg/services/provisioning/dashboards/file_reader_test.go:
```diff
@@ -148,6 +148,47 @@ func TestDashboardFileReader(t *testing.T) {
+
+			Convey("Older dashboard file with different checksum will replace old dashboard", func() {
+				cfg.Options["path"] = oneDashboard
+				absPath, _ := filepath.Abs(oneDashboard + "/dashboard1.json")
+				stat, _ := os.Stat(oneDashboard + "/dashboard1.json")
```
```suggestion
				absPath, err := filepath.Abs(oneDashboard + "/dashboard1.json")
				So(err, ShouldBeNil)
				stat, err := os.Stat(oneDashboard + "/dashboard1.json")
				So(err, ShouldBeNil)
```

cristi- replied:
Thank you!

aknuds1 commented:
Please also merge with master, since it's out of date and there will be build errors.

aknuds1 commented on pkg/services/provisioning/dashboards/file_reader.go:
```diff
@@ -207,8 +207,8 @@ func (fr *FileReader) saveDashboard(path string, folderID int64, fileInfo os.Fil
-	if provisionedData != nil && jsonFile.checkSum == provisionedData.CheckSum {
-		upToDate = true
+	if provisionedData != nil {
+		upToDate = jsonFile.checkSum == provisionedData.CheckSum
```
I think if we should ignore mtime (which I'm unsure about), line 202 should be changed instead, since this change makes the usage of `resolvedFileInfo.ModTime()` ineffective.

wbrowne replied:
Or we update it to

```
upToDate = upToDate || jsonFile.checkSum == provisionedData.CheckSum

cristi- replied:
Won't this change cause the same issue if the file has an older timestamp but a different checksum? The initial time check for `upToDate` will be `true` but the checksum will be different, making the final `upToDate` being `false` again, causing the provisioned dashboard not to be saved in the database.

cristi- replied:
I removed the `mtime` check since it did feel redundant and added a few more tests to catch all 4 possible cases (older/newer file on disk and the same/different checksum).

aknuds1 commented:
The referenced issue needs more info, as I indicated there before we can consider approving this PR, since we need to understand the motivation.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into codespace-1

aknuds1 committed:
Fix tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
I cleaned up the code a bit, for the sake of easy understanding of the changes, and cleanup in the tests (there was a typo among other things).

aknuds1 reviewed (approved):
LGTM! All it changes AFAICT is that only checksum is used to tell if the provisioned dashboard has changed. In my book that's the right thing to do. Thanks for adding tests!

wbrowne reviewed (approved):
Yep looks good to me 👍
